### DISCUSSION TITLE
Renovate seems to freeze on git push. Timeout value needed?

### DISCUSSION THREAD

**Discussion by fullstackzach** [renovatebot/renovate#24792] [Request Help]
Self-hosted Gitlab, Renovate CLI v24.119.4 running in a scheduled gitlab pipeline

**Describe the bug**

Recently in the last week I noticed Renovate seems to freeze on a particular repo. Renovate processes the repo all up until just past git commit, and then freezes completely until the gitlab pipeline reaches its maximum runtime and ends with a failure. Unfortunately that means Renovate doesn't scan the planned repos after this repo either.

I suspect it may be a bug with a new pre-receive git hook that was implemented by my org on the server-side of our git origin. I've seen reports in my org that git pushes have been hanging for developers on this new pre-receive git hook for the https protocol only, and it appears that is what renovate uses for Gitlab.  The suggested workaround for now is to use git with ssh.  I'm not sure how to have Renovate use SSH instead for git.

I'm not certain of the cause yet, and I suspect renovate is not the reason. However, I was wondering if renovate's git.push should implement a reasonable timeout?  I am not certain what the upper bound time of a git.push would be to specify though. I would think best would to be measure at scale and decide given the data. Another option would be to specify a timeout which is very generous.

**Relevant debug logs**

It can be seen here that renovate finishes committing but then the job ends after its maximum runtime of 3hrs. I believe renovate is trying to git.push next based upon reviewing [the code](https://github.com/renovatebot/renovate/blob/main/lib/util/git/index.ts#L687)

<details><summary>Click me to see logs</summary>

```
DEBUG: 2 file(s) to commit (repository=<repo>, branch=<branch>)
3599DEBUG: Committing files to branch <branch> (repository=<repo>, branch=<branch>)
3600DEBUG: git commit (repository=<repo>, branch=<branch>)
3601       "result": {
3602         "author": null,
3603         "branch": "<branch>",
3604         "commit": "<commit sha>",
3605         "root": false,
3606         "summary": {"changes": 2, "insertions": 1143, "deletions": 966}
3607       }
3609ERROR: Job failed: execution took longer than 3h0m0s seconds
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [x] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

I saw SimpleGit does have a [timeout option/plugin](https://github.com/steveukx/git-js/blob/main/docs/PLUGIN-TIMEOUT.md) that might work

**Comment by viceice** [comment]
I see this more a feature than a bug in renovate.

I think we can accept a pr wich adds a timeout option object, which can have different timeout options (we have a lot other timeouts to add there later). Maybe we can simply add the timeout to hostRules and apply to git when applicable.

Another questions is how to configure timeouts for caches? 🤔 

@rarkins what do you think?

**Comment by rarkins** [comment]
We should have a sensible default timeout for `git` commands regardless of making it configurable. Is it something we need to configure `git` with, or the `simple-git` wrapper function with, or handle it directly ourselves? After some Googling, I haven't been able to figure out an obvious way to configure this.

**Comment by rarkins** [comment]
Related: #9708
